core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java

```diff
@@ -83,18 +88,14 @@ private static Block createInternal(int positionCount, Block dictionary, int[] i
         if (dictionary instanceof DictionaryBlock dictionaryBlock) {
             int[] newIds = new int[positionCount];
             for (int position = 0; position < positionCount; position++) {
-                newIds[position] = dictionaryBlock.getId(ids[position]);
+                newIds[position] = dictionaryBlock.getId(ids[idsOffset + position]);
             }
             dictionary = dictionaryBlock.getDictionary();
             dictionarySourceId = randomDictionaryId();
             ids = newIds;
+            idsOffset = 0;
         }
-        return new DictionaryBlock(0, positionCount, dictionary, ids, false, false, dictionarySourceId);
-    }
-
-    DictionaryBlock(int idsOffset, int positionCount, Block dictionary, int[] ids)
-    {
-        this(idsOffset, positionCount, dictionary, ids, false, false, randomDictionaryId());
+        return new DictionaryBlock(idsOffset, positionCount, dictionary, ids, false, false, dictionarySourceId);
     }
 
     private DictionaryBlock(int idsOffset, int positionCount, Block dictionary, int[] ids, boolean dictionaryIsCompacted, boolean isSequentialIds, DictionaryId dictionarySourceId)
```

Comment:
this constructor is used in multiple places inside DictionaryBlok e.g. copyPositions that can make the dictionary to be RLE.
If we really want to protect ourselves from this case we need a precondition inside this constructor

Reply:
ah right, I missed the usages within this class, we need some more changes

Reply:
After looking closer at the other usages, the only way we found to make RLE dictionary out of non-RLE dictionary through get/copyPositions is to use a BlockBuilder as the dictionary. Although it's technically allowed, I'm not sure we need to worry about that. It would be odd to create a DictionaryBlock with a BlockBuilder as the dictionary rather than a pre-built Block.